File: otel-android-mobile/src/main/java/io/opentelemetry/android/mobile/fleet/FleetAlertHandler.kt
package io.opentelemetry.android.mobile.fleet

import android.util.Log
import io.opentelemetry.android.mobile.buffering.MobileLogRecordProcessor
import io.opentelemetry.android.mobile.config.MobileConfig
import io.opentelemetry.android.mobile.config.PrivacyConfig
import java.time.Instant

class FleetAlertHandler(
    private val processor: MobileLogRecordProcessor,
    private val config: MobileConfig,
    private val deduplicator: FleetAlertDeduplicator,
    private val privacyConfig: PrivacyConfig = PrivacyConfig(),
) {
    private val tag = "FleetAlertHandler"
    private val alertTimestamps = mutableListOf<Long>()
    private val maxAlertsPerHour = 5
    private val activeOverrides = mutableMapOf<String, ActiveOverride>()

    fun onFleetAlert(alert: FleetAlert): FleetAlertResult {
        // Validate expiry
        try {
            val expires = Instant.parse(alert.expiresAt)
            if (Instant.now().isAfter(expires)) {
                Log.d(tag, "Fleet alert ${alert.alertId} expired, skipping")
                return FleetAlertResult(alert.alertId, executed = false, reason = "expired")
            }
        } catch (e: Exception) {
            Log.w(tag, "Invalid expiresAt: ${alert.expiresAt}")
            return FleetAlertResult(alert.alertId, executed = false, reason = "invalid_expiry")
        }

        // Dedup
        if (deduplicator.isProcessed(alert.alertId)) {
            Log.d(tag, "Fleet alert ${alert.alertId} already processed, skipping")
            return FleetAlertResult(alert.alertId, executed = false, reason = "duplicate")
        }

        // Rate limit
        if (isRateLimited()) {
            Log.w(tag, "Fleet alert rate limit exceeded")
            return FleetAlertResult(alert.alertId, executed = false, reason = "rate_limited")
        }

        // Execute actions
        val executedActions = mutableListOf<String>()
        val skippedActions = mutableListOf<String>()

        for (action in alert.actions) {
            try {
                val result = executeAction(action, alert.priority)
                if (result) {
                    executedActions.add(action.type)
                } else {
                    skippedActions.add("${action.type}:preempted_or_blocked")
                }
            } catch (e: Exception) {
                Log.e(tag, "Error executing action ${action.type}: ${e.message}")
                skippedActions.add("${action.type}:error")
            }
        }

        deduplicator.markProcessed(alert.alertId)
        recordAlertTimestamp()

        return FleetAlertResult(
            alertId = alert.alertId,
            executed = true,
            actionsExecuted = executedActions,
            actionsSkipped = skippedActions,
        )
    }

    private fun executeAction(action: FleetAction, alertPriority: Int): Boolean {
        return when (action.type) {
            "flush_buffer" -> {
                if (!privacyConfig.allowFleetFlush) return false
                val minutes = action.config["minutes"]?.toIntOrNull() ?: 5
                processor.flushWindow(minutes)
                true
            }
            "set_sampling" -> {
                if (!privacyConfig.allowFleetSampling) return false
                val existing = activeOverrides["sampling"]
                if (existing != null && existing.priority < alertPriority) {
                    return false
                }
                val durationMin = action.config["duration_minutes"]?.toIntOrNull() ?: 10
                activeOverrides["sampling"] = ActiveOverride(
                    key = "sampling", priority = alertPriority, expiresAtMs = System.currentTimeMillis() + durationMin * 60_000L
                )
                true
            }
            "take_screenshot" -> {
                if (!privacyConfig.allowFleetScreenshot) {
                    return false
                }
                true
            }
            else -> {
                Log.w(tag, "Unknown fleet action type: ${action.type}")
                false
            }
        }
    }

    private fun isRateLimited(): Boolean {
        val oneHourAgo = System.currentTimeMillis() - 3_600_000L
        alertTimestamps.removeAll { it < oneHourAgo }
        return alertTimestamps.size >= maxAlertsPerHour
    }

    private fun recordAlertTimestamp() {
        alertTimestamps.add(System.currentTimeMillis())
    }
}

data class FleetAlertResult(
    val alertId: String,
    val executed: Boolean,
    val reason: String? = null,
    val actionsExecuted: List<String> = emptyList(),
    val actionsSkipped: List<String> = emptyList(),
)

data class ActiveOverride(
    val key: String,
    val priority: Int,
    val expiresAtMs: Long,
)

File: otel-ios-mobile/Sources/OTelMobileSDK/Fleet/FleetAlertHandler.swift
/*
 * Copyright 2025 Barry Solomon
 * SPDX-License-Identifier: Apache-2.0
 */
import Foundation

/// Applies incoming `FleetAlert` payloads to local SDK actions:
/// `flush_buffer` (triggers `flushWindow`), `set_sampling` (installs a
/// time-limited active override), `take_screenshot` (gated — no-op today).
///
/// Android parity: mirrors `FleetAlertHandler.kt`. Validates in this order:
/// 1. **Expiry** — `ISO8601DateFormatter` parse of `expiresAt`; past-expiry
///    alerts rejected.
/// 2. **Dedup** — `FleetAlertDeduplicator.isProcessed(alertId)` short-circuits.
/// 3. **Rate limit** — rolling 1-hour window, max 5 accepted alerts.
/// 4. **Privacy gates** — per-action `PrivacyConfig` bool checks.
public final class FleetAlertHandler: @unchecked Sendable {
    public struct Config: Sendable {
        public var maxAlertsPerWindow: Int
        public var rateLimitWindowSeconds: TimeInterval
        /// When true, `set_sampling` alerts may preempt existing overrides
        /// if incoming priority is strictly greater.
        public var honorSetSampling: Bool
        /// When true, `take_screenshot` is allowed (stubbed today — still
        /// gated so privacy posture doesn't drift if/when it ships).
        public var allowScreenshot: Bool
        /// When true, `flush_buffer` alerts are allowed. Off-by-default
        /// deployments can set false to refuse all fleet flushes.
        public var allowFleetFlush: Bool

        public static let `default` = Config(
            maxAlertsPerWindow: 5,
            rateLimitWindowSeconds: 3600,
            honorSetSampling: true,
            allowScreenshot: false,
            allowFleetFlush: true
        )

        public init(
            maxAlertsPerWindow: Int = 5,
            rateLimitWindowSeconds: TimeInterval = 3600,
            honorSetSampling: Bool = true,
            allowScreenshot: Bool = false,
            allowFleetFlush: Bool = true
        ) {
            self.maxAlertsPerWindow = maxAlertsPerWindow
            self.rateLimitWindowSeconds = rateLimitWindowSeconds
            self.honorSetSampling = honorSetSampling
            self.allowScreenshot = allowScreenshot
            self.allowFleetFlush = allowFleetFlush
        }
    }

    public struct SamplingOverride: Sendable, Equatable {
        public let rate: Double
        public let priority: Int
        public let until: Date
    }

    public typealias FlushWindowClosure = @Sendable (UInt64) -> Void

    private let config: Config
    private let dedup: FleetAlertDeduplicator
    private let flushWindow: FlushWindowClosure
    private let lock = NSLock()
    private var alertTimestamps: [Date] = []
    private var currentOverride: SamplingOverride?

    public init(
        config: Config = .default,
        dedup: FleetAlertDeduplicator = FleetAlertDeduplicator(),
        flushWindow: @escaping FlushWindowClosure
    ) {
        self.config = config
        self.dedup = dedup
        self.flushWindow = flushWindow
    }

    public func handle(_ alert: FleetAlert) -> FleetAlertResult {
        if let reason = validateExpiry(alert) {
            return FleetAlertResult(alertId: alert.alertId, accepted: false, reason: reason)
        }
        if dedup.isProcessed(alert.alertId) {
            return FleetAlertResult(alertId: alert.alertId, accepted: false, reason: "duplicate")
        }
        if let reason = rateLimitCheck() {
            return FleetAlertResult(alertId: alert.alertId, accepted: false, reason: reason)
        }

        var executed: [String] = []
        var skipped: [String] = []

        for action in alert.actions {
            switch action.type {
            case "flush_buffer":
                if !config.allowFleetFlush {
                    skipped.append("flush_buffer:privacy_gate")
                    continue
                }
                let minutes = UInt64(action.config["minutes"].flatMap { Int($0) } ?? 5)
                flushWindow(minutes)
                executed.append("flush_buffer")
            case "set_sampling":
                if !config.honorSetSampling {
                    skipped.append("set_sampling:privacy_gate")
                    continue
                }
                guard let rateStr = action.config["rate"], let rate = Double(rateStr) else {
                    skipped.append("set_sampling:missing_rate")
                    continue
                }
                let durationSeconds = Double(action.config["duration_seconds"] ?? "300") ?? 300
                let candidate = SamplingOverride(
                    rate: rate,
                    priority: alert.priority,
                    until: Date().addingTimeInterval(durationSeconds)
                )
                if let existing = currentOverride, Date() < existing.until,
                   candidate.priority <= existing.priority {
                    skipped.append("set_sampling:lower_priority")
                    continue
                }
                lock.lock()
                currentOverride = candidate
                lock.unlock()
                executed.append("set_sampling")
            case "take_screenshot":
                if !config.allowScreenshot {
                    skipped.append("take_screenshot:privacy_gate")
                    continue
                }
                // Stubbed — iOS screenshot module is gated behind the
                // privacy design. When it ships, wire here.
                skipped.append("take_screenshot:not_implemented")
            default:
                skipped.append("\(action.type):unknown")
            }
        }

        lock.lock()
        alertTimestamps.append(Date())
        lock.unlock()
        dedup.markProcessed(alert.alertId)

        return FleetAlertResult(
            alertId: alert.alertId,
            accepted: true,
            reason: nil,
            actionsExecuted: executed,
            actionsSkipped: skipped
        )
    }

    /// Current sampling override, if any and still valid. Returns nil when
    /// the override has expired (without mutating state — expired
    /// overrides remain until the next `handle(_:)` arrives with a new
    /// priority-wins check).
    public func activeSamplingOverride() -> SamplingOverride? {
        lock.lock(); defer { lock.unlock() }
        guard let ov = currentOverride, Date() < ov.until else { return nil }
        return ov
    }

    // MARK: - Validation

    private func validateExpiry(_ alert: FleetAlert) -> String? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let parsed = formatter.date(from: alert.expiresAt)
            ?? ISO8601DateFormatter().date(from: alert.expiresAt)
        guard let expiresAt = parsed else { return "invalid_expires_at" }
        if expiresAt <= Date() { return "expired" }
        return nil
    }

    private func rateLimitCheck() -> String? {
        lock.lock(); defer { lock.unlock() }
        let cutoff = Date().addingTimeInterval(-config.rateLimitWindowSeconds)
        alertTimestamps.removeAll { $0 < cutoff }
        if alertTimestamps.count >= config.maxAlertsPerWindow {
            return "rate_limited"
        }
        return nil
    }
}
